vijan - Technical Training

VincentJanssen-Code opened the pull request:
This pull request is for the [Server 101 tutorial](https://www.odoo.com/documentation/19.0/developer/tutorials/server_framework_101.html) of the on boarding of vijan

VincentJanssen-Code committed:
[IMP] estate: Server101 Finished

[IMP] estate: Server101 Finished Chapter 3

[IMP] estate: Server101 Finished Chapter 4

[IMP] estate: Server101 Finished Chapter 5

[IMP] estate: Server101 Finished Chapter 6

[IMP] estate: Server101 Finished Chapter 7

[IMP] estate: Server101 Finished Chapter 8

[IMP] estate: Server101 Finished Chapter 9

[IMP] estate: Server101 Finished Chapter 10

[IMP] estate: Server101 Finished Chapter 10

[IMP] estate: Server101 WIP Chapter 11

[CLN] estate: code cleanup to match Ruff

[CLN] estate: Server101 Cleanup

[CLN] estate: Server101 Cleanup final

[CLN] estate: Server101 Appling PR requested changes and cleanup of the code for readability

[FIX] estate:  Server101 Fixing  bugs based on cleanup

[IMP] estate: Server101 Finished Chapter 11

[IMP] estate : Server101 Finished Chapter 12

[FIX] estate: Fixed bug where the stat button from property type to property offer would crash on new install of the estate module

[IMP]  estate: Server101 Finished Chapter 13

[IMP] estate: Server101 Finshed Chapter 14

[FIX] estate: Added  error message translation useing self.env._() and applied fix in the PR

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1055.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1055)

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,84 @@
+    _positif_expected_price = models.Constraint("CHECK (expected_price > 0)","A price can't be negatif");
+    _positif_selling_price = models.Constraint("CHECK (selling_price > 0)","A price can't be negatif");
+
+    state = fields.Selection(selection = [("New","New"), ("Offer_Received","Offer Received") ,("Offer_Accepted","Offer Accepted"), ("Sold","Sold"), ("Cancelled","Cancelled")])
```
for the selection fields I prefer to write them that way to make the choices clear while we look at the field. and for the values the first part of each tuple is the technical value for the field which should be written like a var name in snake case 🐍 
```suggestion
    state = fields.Selection([
        ("new","New"),
        ("offer_received","Offer Received"),
        ("offer_accepted","Offer Accepted"),
        ("sold","Sold"),
        ("cancelled","Cancelled")
    ], default="new")
```

amah-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from . import estate_property
+from . import estate_property_type
+from . import estate_property_tag
+from . import estate_property_offer
```
newline missed

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,84 @@
+    name = fields.Char(required=True,default="Unkown")
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date("Available From",copy=False,default= fields.Datetime.today() + relativedelta(months=3))
```
when you run the server for the first time the default values will be stored on the db so when you open the next day it will be 3 months from day it got evaluated. but with the lambda it will be expression that will be evaluated every time 
```suggestion
    date_availability = fields.Date("Available From",copy=False,default= lambda self: fields.Datetime.today() + relativedelta(months=3))
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,84 @@
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date("Available From",copy=False,default= fields.Datetime.today() + relativedelta(months=3))
+    last_seen= fields.Date("Last Seen", default=fields.Datetime.now)
```
same as `date_availability`

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,84 @@
+    last_seen= fields.Date("Last Seen", default=fields.Datetime.now)
+    expected_price = fields.Float(required=True)
+    selling_price = fields.Float(readonly=True,copy=False)
+    best_price = fields.Float(string="Best Price",compute="_get_best_price")
```
we tend to use the `_compute` prefix for all the compute methods 👀 
```suggestion
    best_price = fields.Float(string="Best Price",compute="_compute_best_price")
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    _positif_expected_price = models.Constraint("CHECK (expected_price > 0)", "A price can't be negatif")
+    _positif_selling_price = models.Constraint("CHECK (selling_price > 0)", "A price can't be negatif")
+
+    state = fields.Selection(selection=[("New", "New"), ("Offer_Received", "Offer Received"), ("Offer_Accepted", "Offer Accepted"), ("Sold", "Sold"), ("Cancelled", "Cancelled")])
```
I prefer to write Selection that way to make the choices clear while reading the file. Also the tuples' first value should be written in snake_case 🐍 as they are technical names
```suggestion
    state = fields.Selection(selection=[
        ("new", "New"),
        ("offer_received", "Offer Received"),
        ("offer_accepted", "Offer Accepted"),
        ("sold", "Sold"),
        ("cancelled", "Cancelled")
    ], default="new")
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    name = fields.Char(required=True, default="Unkown")
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date("Available From", copy=False, default=fields.Datetime.today() + relativedelta(months=3))
```
without the lambda this will be evaluated once the server is running and will not be re-evaluated with each new record which is not the case we want 🤔 
```suggestion
    date_availability = fields.Date("Available From", copy=False, default=lambda self: fields.Datetime.today() + relativedelta(months=3))
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date("Available From", copy=False, default=fields.Datetime.today() + relativedelta(months=3))
+    last_seen = fields.Date("Last Seen", default=fields.Datetime.now)
```
same here as `date_availability` 👀

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    garage = fields.Boolean()
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(selection=[('North', 'North'), ('South', 'South'), ('East', 'East'), ('West', 'West')])
```
same here as `state`

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    last_seen = fields.Date("Last Seen", default=fields.Datetime.now)
+    expected_price = fields.Float(required=True)
+    selling_price = fields.Float(readonly=True, copy=False)
+    best_price = fields.Float(string="Best Price", compute="_get_best_price")
```
we tend to use '_compute' prefix for the compute methods

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    garden = fields.Boolean()
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection(selection=[('North', 'North'), ('South', 'South'), ('East', 'East'), ('West', 'West')])
+    total_area = fields.Float(compute="_get_total_area", string="Total Area")
```
same as `best_price`

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    def sell_property(self):
+        for record in self:
+            if (record.state == "Cancelled"):
+                raise exceptions.UserError("Can't sell cancelled property.")
```
you could import it like the `ValidationError` to not import the whole exceptions module

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+            for offer in record.offer_ids:
+                if (offer.price >= max):
+                    max = offer.price
+                record.best_price = max
```
the `.mapped('field_name')` method return a list with all the values for this field for this record set
```suggestion
            record.best_price = max(record.offer_ids.mapped('price'))
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+    def _get_best_price(self):
+        for record in self:
+            max: int = 0
+            if (len(record.offer_ids) == 0):
```
the empty record set is a false value so you can just check for that 
```suggestion
            if not record.offer_ids:
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,83 @@
+                record.best_price = max
+
+    @api.onchange("garden")
+    def _garden_pre_fill(self):
```
we have the `_on_change` prefix also 👀 
```suggestion
    def _on_change_garden(self):
```

amah-odoo reviewed (commented):
Huge Work 👏🏼 

let's go for those today... TO BE CONTINUED 👀

amah-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_estate_property_model,access_estate_property_model,model_estate_property,base.group_user,1,1,1,1
+access_estate_property_type_model,access_estate_property_type_model,model_estate_property_type,base.group_user,1,1,1,1
+access_estate_property_tag_model,access_estate_property_tag_model,model_estate_property_tag,base.group_user,1,1,1,1
+access_estate_property_offer_model,access_estate_property_offer_model,model_estate_property_offer,base.group_user,1,1,1,1
```
missing end of file newline 👀

amah-odoo commented on estate/views/estate_menu_views.xml:
```diff
@@ -0,0 +1,12 @@
+            <menuitem id="estate_property_tag_menu_acion" action="estate_property_tag_model_action"/>
+        </menuitem>
+    </menuitem>
+</odoo>
```
missing end of file newline 👀

amah-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,59 @@
+                record.date_deadline = fields.Datetime.now() + relativedelta(days=record.validity)
+                return
+            record.date_deadline = record.create_date + relativedelta(days=record.validity)
+
+    def _inverse_date_deadline(self):
+        for record in self:
+            record.validity = (datetime.combine(record.date_deadline, time()) - record.create_date).days
+
+    @api.model
+    def create(self, vals_list):
+        for record in vals_list:
+            property = self.env['estate.property'].browse(record['property_id'])
+            if (float_compare(record['price'], property.best_price, 2) == -1):
+                raise UserError("This offer is lower than what has already been offered.")
+            property.state = 'offer_received'
+        return super().create(vals_list)
```
I think we can optimize this browse and write calls for the property model by getting all the properties before the loop and update all of them once after the creation. Also I think we should update the state at least after the offer creation or at the property model it self 🤔

amah-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,45 @@
+        </field>
+    </record>
+
+</odoo>
```
missing end of file newline 👀

amah-odoo commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,34 @@
+        <field name="view_mode">list,form</field>
+    </record>
+
+    <record id="estate_property_tag_model_form" model="ir.ui.view">
```
I think we do not need the form and the list as it is only the name and this is the default for list 🤔

amah-odoo reviewed (commented):
Huge work 💯 

with the translation this will be more than prefect module

VincentJanssen-Code committed:
[IMP] awesome_owl: Web Framework Finished Chapter 1

[IMP] awesome_owl: Web Framework Finished Chapter 1 Part 9-14

VincentJanssen-Code committed:
[IMP] awesome_owl: Web Framework Finished Chapter 2

[FIX] awesome_owl: Invisible dialog box

[FIX] awesome_owl:  Dummy push

[IMP]  awsome_owl: Framework Finished Chapter 7-12

VincentJanssen-Code committed:
[IMP] awesome_clicker: Web Framework Master Chapter 1 Finish

[IMP] awesome_clicker: Web Framework Master Chapter 1 Finished Part 10-21

VincentJanssen-Code committed:
[IMP] estate : Added Restrict access to data from the documentation's tutorial

VincentJanssen-Code committed:
[FIX] Bug with querry generation on tests
